packages/core/src/new-api/internal/execution/chain-dispatcher.ts

```diff
@@ -0,0 +1,160 @@
+import { Contract, ContractFactory, ethers } from "ethers";
+
+import { Adapters } from "../../types/adapters";
+import { ArgumentType } from "../../types/module";
+import { ChainDispatcher } from "../types/chain-dispatcher";
+import { assertIgnitionInvariant } from "../utils/assertions";
+
+import { AccountsState } from "./execution-engine";
+
+export class ChainDispatcherImpl implements ChainDispatcher {
+  private _accountsState: AccountsState;
+
+  constructor(private _adapters: Adapters) {
+    this._accountsState = {};
+  }
+
+  // TODO: should nonce management be separated out into its own
+  // class?
+  public async allocateNextNonceForAccount(address: string): Promise<number> {
```

Comment:
This may be complecting nonce management with the ethers chain interactions, we should consider moving it out.

In general the nonce initialization check still needs done.

Reply:
We need to resolve nonces from history when we are restarting.

This might be the place to detect user interference with the nonce.

Reply:
It needs to get the pending transaction count.